### PR TITLE
fix(auth): Unify ZMI, HTML form, and API login

### PR DESCRIPTION
**Pull request by rpatterson**
Fix some behavior in the PAS plugin that treated API login requests differently than
classic HTML form login requests, namely that the POST'ed login credentials weren't
extracted and authenticated like they are when the classic HTML login form is POST'ed.
IMO, the root cause here was a minor case of not actually fulfilling the PAS plugin
interface.  In the case of the classic HTML login form, it's another PAS plugin that
extracts the POST'ed credentials, but since the API login POST submits JSON, it has to
be the JWT plugin that extracts them.

Also reproduce the same handling that the classic Plone HTML login form does after login
has succeeded.  This includes setting the classic HTML login form cookie so that
authenticating to the API also authenticates the user to the classic HTML UI (as well as
the ZMI when the user has the `Manager` role).  But it also includes other core Plone
login behavior.

In the case of Volto, there's one more edge case remaining.  Namely, if the browser is
already authenticated such as by having previously submitted the classic HTML login
form, a fresh load of the Volto UI will still show them as logged out even though their
API requests succeed as authenticated.  I assume this is a problem in the React
component state that doesn't do an actual API check to see if the user is already
authenticated and if I'm right a fix for that can't be included here.

----

This is my first Volto-adjacent PR.  Please LMK what parts of etiquette I may be unaware
of.  Also please pick all the nits possible so I can learn how to better form PRs for
the Volto project going forward.  IOW, *let me have it*!  ;-)

### CODE REVIEW

**Comment by rpatterson** [comment]
@jenkins-plone-org please run jobs

File: src/plone/restapi/services/auth/login.py
Comment (rpatterson) [reply src/plone/restapi/services/auth/login.py:71]: I'm old, so checks out.  ;-)  I'm certainly open to suggestions.  I debated this myself when writing that line.  I opted for this because if a given deployment/application overrides that view I figured using any modified logic from the override would be less likely to be surprising.  Is there a more "modern" way to lookup a view in a context succinctly?  Is using the low-level ZCA adapter lookup considered  better?  (I'd find that hard to swallow personally.)  OTOH, the `._post_login()` method being used is a soft "private" method per Python conventions so I could appreciate an argument based on that except that what we're trying to accomplish here requires either calling that "private" method or violating DRY.

I'd love to hear any further thoughts.

Comment (tiberiuichim) [reply src/plone/restapi/services/auth/login.py:71]: On the other hand, the title of the PR is "unify html form and API login". Can't have one without the other :)

Comment (sneridagh) [reply src/plone/restapi/services/auth/login.py:71]: Although we opted for not using methods in browser views in most of the parts of plone.restapi it's true that we are use them anyways in some others. The rationale is to not to depend in them, specially if at some point in the future we get rid of those browser views (eg. moving out of the core the packages that include them, in an hypothetical Plone core "lite" with no Classic UI packages).

Comment (tiberiuichim) [reply src/plone/restapi/services/auth/login.py:71]: I guess I'd feel more comfortable with `getMultiAdapter((self.context, self.request), name="login")`

File: src/plone/restapi/tests/test_functional_auth.py
Comment (sneridagh) [inline src/plone/restapi/tests/test_functional_auth.py:2]: 100% sure you already know the Plone convention, could be that a wrong import sort linter was applied here?

**Review by sneridagh — COMMENTED** [review]
Overall LGTM, curious about the use case that triggered this.

@alecghica the feature you told me several weeks ago in the Volto Team meeting was different, right?

@buchi @lukasgraf @tisto It would be great if you could also take a look! Thanks!

File: src/plone/restapi/tests/test_functional_auth.py
Comment (rpatterson) [reply src/plone/restapi/tests/test_functional_auth.py:2]: Well in that case you're 100% wrong about what I already know.  ;-)  I am not aware of a Plone convention on sorting imports.  I *am* aware of some portions of the old Zope community that preferred sorted imports.  I am also aware that I do not prefer sorted imports, at least not without some division demonstrating the level in the stack, e.g.: standard library imports, framework agnostic non-standard-library imports, framework imports, imports from this add-on.

Regardless, I'm happy to follow a community convention on imports.  If this is a strong community standard, shouldn't this be in the CI checks?  Failing that, I've looked at a few other modules and it's not at all clear from those what the convention is.  So if CI isn't telling me what to change, care to give me link, @sneridagh, or otherwise help me understand what change you would prefer I make here?  Thanks!

Comment (tisto) [reply src/plone/restapi/tests/test_functional_auth.py:2]: It is kind of best practice in the Plone community to move the "import *" statements below the other imports. We use this convention in Plone core packages and in plone.restapi. Therefore it seems indeed odd to move them to the top here.

There is/was also some initiative to automatically sort imports and make them checkable on CI, which would be indeed the way to go. Though, that never was fully enforced. Partly because it is not sufficient to just let the CI check those things, you need proper pre-commit hooks, IDE integration, etc.

That being said, you are 100% correct and this is indeed a problem for people that do not work with Plone core code on a daily basis. Though I'd suggest just not move around the "import *" to have at least a little bit of consistency in the package.

File: src/plone/restapi/services/auth/login.py
Comment (rpatterson) [reply src/plone/restapi/services/auth/login.py:71]: > I guess I'd feel more comfortable with `getMultiAdapter((self.context, self.request), name="login")`

I was never comfortable with this ZCA usage pattern as it doesn't seem explicit.  Couldn't there be other adapters registered for 2 objects that have the name `login` that aren't anything resembling a view?  I always felt that an adapter lookup should always specify the most specific provided interface that's appropriate for the use case.

That said, your suggestion *is* the lookup that ultimately happens in the `./plone.restapi/buildout.cfg` code.  Before that, however, `OFS.Traversable.Traversable.unrestrictedTraverse()` does a namespace lookup on the `@@` prefix which uses a `queryMultiAdapter((object, request), ITraversable, ns)` to get the actual code that actually performs the ultimate `queryMultiAdapter((self.context, self.request), name=name)` lookup to get the actual view.  So if a deployment/application has registered an alternate namespace traverser then the short-circuit straight to `getMultiAdapter((self.context, self.request), name="login")` in this PR would not use that alternate traverser which I think would be more surprising than the current implementation.

Having said all that, I don't actually care that much.  ;-)  Just repeat the request and I'll happily change to the suggested `queryMultiAdapter((self.context, self.request), name=name)`.

Comment (tisto) [reply src/plone/restapi/services/auth/login.py:71]: There is no obvious solution to this IMHO, and all points raised here are valid claims. Using getMultiAdapter does sound like a more modern approach than unrestrictedTraverse. Depending on internal methods is a code smell indeed. Though, duplicating the code in the restapi code base also has its downsides (e.g. when the code changes in one of the sides). Bottom line is that if we remove browser views in the future, we will have to refactor things in restapi anyways I guess and in the meantime duplicating code has the potential to cause problems. Therefore I tend to agree with @rpatterson's tradeoff here and I'd just move to getMultiAdapter and add a few lines of explanation.

**Comment by mamico** [comment]
> In the case of Volto, there's one more edge case remaining. Namely, if the browser is
> already authenticated such as by having previously submitted the classic HTML login
> form, a fresh load of the Volto UI will still show them as logged out even though their
> API requests succeed as authenticated. I assume this is a problem in the React
> component state that doesn't do an actual API check to see if the user is already
> authenticated and if I'm right a fix for that can't be included here.

@rpatterson,  time ago I started with https://github.com/plone/plone.restapi/tree/mamico/ext-login that probably  could
cover the above use case and different other use cases, like external authentication method (e.g. pas.plugin.authomatic).
Any hints would be appreciated.

**Comment by alecghica** [comment]
> Overall LGTM, curious about the use case that triggered this.
> 
> @alecghica the feature you told me several weeks ago in the Volto Team meeting was different, right?

It's related to that.

File: src/plone/restapi/tests/test_functional_auth.py
Comment (rpatterson) [reply src/plone/restapi/tests/test_functional_auth.py:2]: K, thanks much for the explanation!  My default mode is to try to clean up things I touch.  I'll try to remember not to do that with imports in Plone core going forward.  I'll also revert the import order here to the way it was before.

**Comment by rpatterson** [comment]
I believe I've addressed all the feedback, @tiberiuichim @tisto.  LMK if I missed anything.  Otherwise,  what are the next steps?  I assume a review approval.  After that, for this project is the merge button something I push?  Any follow up steps expected of me?

**Comment by rpatterson** [comment]
@jenkins-plone-org please run jobs

**Comment by rpatterson** [comment]
> @rpatterson, time ago I started with https://github.com/plone/plone.restapi/tree/mamico/ext-login that probably could
> cover the above use case and different other use cases, like external authentication method (e.g. pas.plugin.authomatic).
> Any hints would be appreciated.

Thanks for reaching out, @mamico!  First, a minor nit pick, it looks like your 2nd commit is just a line ending change for the 1st commit so I assume they should be squashed together?

Is the `auth_token` cookie already handled in the Volto auth components?  I thought Volto was using the `Authorization` header?  Does it use both?  If you happen to know and have the time to educate me, that would be a favor.

Proceeding from there assuming that setting that cookie has the desired effect in the Volto components, then I would say your event subscriber code should be converted to a `plone.restapi.pas.plugin:JWTAuthenticationPlugin.updateCredentials(...)` method.  If done there then portals can use PAS configuration to opt in/out of this behavior like they do for other PAS plugins.  See [`Products.PluggableAuthService.plugins.CookieAuthHelper:CookieAuthHelper.updateCredentialsupdateCredentials(...)`](https://github.com/zopefoundation/Products.PluggableAuthService/blob/master/src/Products/PluggableAuthService/plugins/CookieAuthHelper.py#L175-L179) for an example.  If done that way, I would describe what your code is trying to do would make the JWT PAS plugin a more complete implementation of the PAS plugin interface.  IOW, sounds like an improvement to me.

I haven't thought about your code in depth as I would for a PR review so just request a review from me in a PR if you'd like a more detailed evaluation.  One potential issue jumps out at me, how would your code set the same JWT token in both the cookie and return it in the JSON  response to a login `POST`?

**Comment by tisto** [comment]
@rpatterson first of all, thanks for your PR and the time you put into discussing our nitpicking comments. The PR looks good to me. I'd just like to give it a shot locally and maybe @buchi and/or @lukasgraf to have a second look. Then we are good to merge and I can do a release right away. There is no further action from your side required I guess. 

@buchi @lukasgraf let us know if you want this fix in plone.restapi 7.x. Since we started the new Plone 5.2/Python3 branch (8.x.x) we have to decide if we want this fix there as well.

**Comment by rpatterson** [comment]
> ...There is no further action from your side required I guess.

K, works for me.  ;-)  Quick caveat, I just tuned into [the CI failure from the run I kicked off after my latest push](https://jenkins.plone.org/job/pull-request-6.0-3.8/191/testReport/junit/Products.CMFPlone.tests.test_robot/RobotTestCase/Scenario_Thememapper_LESS_builder/).  The output doesn't say much to me so I can't even argue whether or not it's related to my changes.  I've never gotten to fluency with Robot tests so it may take me time to understand this failure.  OTOH, I've wanted to achieve fluency there so if someone thinks this is related, give me some hints and I'll get to it!  Otherwise I'll assume this PR is in others hands until I hear otherwise.

**Comment by alecghica** [comment]
> let us know if you want this fix in plone.restapi 7.x.

@tisto would be great to have it for plone.restapi 7.x.

**Comment by wesleybl** [comment]
> I can't even argue whether or not it's related to my changes.

@rpatterson It's common for some robot tests to fail on Jenkins unrelated to your PR. When this happens, you have to restart the job that failed. To do this, access the job that failed, click on "Login" to authenticate, click on the menu "job number" -> Retry

**Comment by mamico** [comment]
@rpatterson thanks,

> Is the `auth_token` cookie already handled in the Volto auth components? I thought Volto was using the `Authorization` header? Does it use both? If you happen to know and have the time to educate me, that would be a favor.

AFAIK, Volto uses the token in the `Authorization` header when calling the restapi backend, but the token is actually stored in a browser session cookie (https://github.com/plone/volto/blob/3772e3c971c3eb5e7cdde0bf072f5e660f41fc3c/src/helpers/AuthToken/AuthToken.js#L54). 
I'm using this approach in a working in progress pas plugins and ... it seems to works fine (https://github.com/mamico/pas.plugins.oidc/blob/main/src/pas/plugins/oidc/plugins.py#L123). With this implementation in p.restapi, some code there could be removed.

> Proceeding from there assuming that setting that cookie has the desired effect in the Volto components, then I would say your event subscriber code should be converted to a `plone.restapi.pas.plugin:JWTAuthenticationPlugin.updateCredentials(...)`..

Good point, I'll try to think about it... 

> One potential issue jumps out at me, how would your code set the same JWT token in both the cookie and return it in the JSON response to a login `POST`?

Sorry, I don't understand this last point ... the main picture is: whichever way you authenticate with a legacy plone pas plugin, if correctly the IUserLoggedInEvent event will be called, this code will return to the user a valid auth_token cookie, this is all Volto needs for subsequent calls to the backend.

**Comment by rpatterson** [comment]
> > I can't even argue whether or not it's related to my changes.
> 
> @rpatterson It's common for some robot tests to fail on Jenkins unrelated to your PR. When this happens, you have to restart the job that failed. To do this, access the job that failed, click on "Login" to authenticate, click on the menu "job number" -> Retry

Thanks!  I didn't know I could log into Plone Jenkins.  I've restarted it.

To put in my $0.02 that no one asked for, I think an intermittent test is often worse than no test.  It frustrates the community of developers and trains them to ignore tests over time.  We want our tests to help developers fall in love with testing.  This goes double for CI.  Does Robot Framework have an equivalent of Python's `unittest.TestCase.skipTest(...)`?

**Comment by rpatterson** [comment]
> Sorry, I don't understand this last point ...

To clarify, @mamico.  Doesn't the login API endpoint view from `plone.restapi` call `plugin.create_token(...)` as well and retrun that one in the JSON response?  Woudn't that be a different token from the one you set in the cookie?

**Comment by mamico** [comment]
> > Sorry, I don't understand this last point ...
> 
> To clarify, @mamico. Doesn't the login API endpoint view from `plone.restapi` call `plugin.create_token(...)` as well and retrun that one in the JSON response? Woudn't that be a different token from the one you set in the cookie?

@rpatterson. It's the same token, in fact I use the same `plugin.create_token(...)` (https://github.com/plone/plone.restapi/blob/mamico/ext-login/src/plone/restapi/subscribers.py#L25) as the login API. The only difference is that, using the Volto form login, Volto (js on the browser) call the login API and set the token (in the JSON response) in a cookie for persistence. Using the "legacy" Plone login the code that I propose set, in an HTTP response, the same type of token in the same cookie, and Volto next use it.

**Comment by rpatterson** [comment]
Apologies for the back and forth, @mamico.  Maybe you could create a PR, label it `WIP` and we can continue discussion there?

> @rpatterson. It's the same token, in fact I use the same `plugin.create_token(...)` (https://github.com/plone/plone.restapi/blob/mamico/ext-login/src/plone/restapi/subscribers.py#L25) as the login API.

Yeah, I saw that, but I thought when I read the `def create_token(...):` method code that it created a new, different token for each call.  Did I read that code wrong?  Very possible.  ;-)

**Comment by rpatterson** [comment]
Just a quick reminder, is anyone able to review? (CC: @tisto, @buchi, @sneridagh, @lukasgraf)

**Review by sneridagh — APPROVED** [review]

**Comment by rpatterson** [comment]
@jenkins-plone-org please run jobs

**Comment by tisto** [comment]
@tiberiuichim this is exactly what @sneridagh did if I am not mistaken. This did not solve the issue though. :/

**Comment by rpatterson** [comment]
I think I've resolved all the outstanding issues brought up here in #1303, @avoinea @tisto @tiberiuichim.  LMK if I'm missing something!

**Pull request merged**
